awscli/customizations/history/db.py

```diff
@@ -0,0 +1,246 @@
+        event_type = record['event_type']
+        json_serialized_payload = json.dumps(record['payload'],
+                                             cls=PayloadSerializer)
+        db_record = (
+            record['command_id'],
+            record.get('request_id'),
+            record['source'],
+            event_type,
+            record['timestamp'],
+            json_serialized_payload
+        )
+        return db_record
+
+
+class DatabaseRecordReader(object):
+    _ORDERING = 'ORDER BY timestamp'
+    _GET_LAST_ID_RECORDS = """
+    SELECT * FROM records
+    WHERE id =
+    (SELECT id FROM records WHERE timestamp =
+    (SELECT max(timestamp) FROM records)) %s;""" % _ORDERING
+    _GET_RECORDS_BY_ID = 'SELECT * from records where id = ? %s' % _ORDERING
+
+    def __init__(self, connection=None):
+        if connection is None:
+            connection = DatabaseConnection()
+        self._connection = connection
+        self._connection.row_factory = self._row_factory
+
+    def _row_factory(self, cursor, row):
+        d = {}
+        for idx, col in enumerate(cursor.description):
+            val = row[idx]
+            if col[0] == 'payload':
+                val = json.loads(val)
+            d[col[0]] = val
+        return d
+
+    def iter_latest_records(self):
+        cursor = self._connection.cursor()
+        cursor.execute(self._GET_LAST_ID_RECORDS)
+        for row in cursor:
+            yield row
+
+    def iter_records(self, record_id):
+        cursor = self._connection.cursor()
+        cursor.execute(self._GET_RECORDS_BY_ID, [record_id])
+        for row in cursor:
+            yield row
+
+
+class RecordBuilder(object):
+    _REQUEST_LIFECYCLE_EVENTS = set(
+        ['API_CALL', 'HTTP_REQUEST', 'HTTP_RESPONSE', 'PARSED_RESPONSE'])
+    _START_OF_REQUEST_LIFECYCLE_EVENT = 'API_CALL'
+    _BYTES_BODY_PAYLOADS = set(['HTTP_REQUEST', 'HTTP_RESPONSE'])
+
+    def __init__(self):
+        self._identifier = None
+        self._locals = threading.local()
+
+    def _get_current_thread_request_id(self):
+        request_id = getattr(self._locals, 'request_id', None)
+        return request_id
+
+    def _start_http_lifecycle(self):
+        setattr(self._locals, 'request_id', str(uuid.uuid4()))
+
+    def _get_request_id(self, event_type):
+        if event_type == self._START_OF_REQUEST_LIFECYCLE_EVENT:
+            self._start_http_lifecycle()
+        if event_type in self._REQUEST_LIFECYCLE_EVENTS:
+            request_id = self._get_current_thread_request_id()
+            return request_id
+        return None
+
+    def _format_payload(self, event_type, payload):
+        # If the payload has a body key it can be in bytes, so it needs to be
+        # converted to utf-8 if possible so it can be serialized as JSON later.
+        if event_type in self._BYTES_BODY_PAYLOADS:
+            body = payload.get('body')
+            if body is not None:
+                payload['body'] = ensure_text_type(body)
```

Comment:
I don't think you answered my original question about this (I might have missed it though).  What specific instances are you running into for this?  Additionally, if a payload can be bytes, then it's possible that you can't utf-8 encode it.  It looks like you just removed the previous version's base64 encoding.  Are you certain we don't need it?

Reply:
The body key on http_request and http_response payloads is as far as I can tell always bytes if it is not None.